Add a bot that sends fake reviews

cravtos opened the pull request.

cravtos committed:
Add a bot that sends fake reviews

cravtos committed:
Minor cleanup

* UsE bEttErEr EnGliSh
* Remove redundant 'continue' in switch statement
* Explicitly return nil error at function end

cravtos committed:
Increment reviewsDone in for-loop

manmolecular commented:
Привет! Спасибо за PR. Сделаю ревью как только появится возможность.

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,202 @@
+func genReview() string {
+	pronouns := []string{"I", "My family", "My wife", "My dog"}
+	verbs := []string{"tried out", "bought", "used", "utilized"}
+	duration := []string{"1 month", "2 days", "1 year"}
```
Комментогенерация пушка 👍🏻

manmolecular replied:
Я думаю, это лучше было бы вынести в var в начало скрипта, нет смысла создавать эти массивы каждый раз при вызове `getReview()`:

```go
var (
	pronouns = []string{"I", "My family", "My wife", "My dog"}
	verbs = []string{"tried out", "bought", "used", "utilized"}
	duration = []string{"1 month", "2 days", "1 year"}
)
```

cravtos replied:
Вынес (48e3ee3)

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,202 @@
+
+	f := pronouns[rand.Intn(len(pronouns))]
+	s := verbs[rand.Intn(len(verbs))]
+	t := duration[rand.Intn(len(duration))]
```
Не очень удачный нэйминг, сходу непонятно что есть f, s, t - лучше хотя бы слово написать осмысленное.

cravtos replied:
Поправил (4284a7c)

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,202 @@
+}
+
+func genAuthor() string {
+	names := []string{"Petya", "Sasha", "Grisha", "Misha", "Oleg"}
```
Я думаю, names можно вынести как var в начало скрипта

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,202 @@
+	values.Set("comment_post_ID", postID)
+	values.Set("comment_parent", "0")
+
+	// NOTE: program can be parallelized if proxies are used
```
К этой заметке: а что если здесь пробовать что-то в стиле password spraying? Идея в том, что если приложение начинает нас стопить из-за большого числа ревью на один товар, то нам можно пробовать делать наоборот - фиксировать коммент и рассылать его на разные товары. 
Т.е., сейчас товар - константа, а ревью - переменная. А мы могли бы сделать наоборот, ревью - константа, а продукт постоянно меняется. 

https://owasp.org/www-community/attacks/Password_Spraying_Attack

Но это не руководство к действию или написанию, просто скорее как вариант байпаса - круговой обход товаров.

cravtos replied:
Это было/есть идея для второго бота, уже для спама по всем товарам)

cravtos replied:
Добавил поддержку прокси. Теперь на каждый прокси запускается свой бот: 4b00165

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,202 @@
+}
+
+func leaveReview(host, postID string, rating int) error {
+	to := fmt.Sprintf("http://%s/wp-comments-post.php", host)
```
Схема (http) я думаю не должна быть зашита - вдруг мы захотим применить скрипт на другом магазе, или прикрутим себе https, и т.д. и т.п., я думаю это должно браться из URL который передается в аргументе

cravtos replied:
Начал брать схему из переданной ссылки (84afb80). 

Если схемы нет, то бот падает с ошибкой, что схема не поддерживается. Так и оставить или лучше брать http:// за дефолт?

manmolecular replied:
Я думаю, чтобы минимизировать "схемы нет" ситуацию, нужно указать в help, что ссылка должна быть полного вида. Но, по природе своей, я не думаю что кто-то будет вводить ссылку руками (т.е. не скопирует её из браузера), т.е. вероятность потерять схему _крайне мала_ но конечно же возможна.

Поэтому, по дефолту в современных интернетах везде лучше всего ставить **https://**, очевидно, НО конкретно в нашем случае, т.к. наш бот поднимается без сертов на **http://** везде, мы будем по дефолту юзать **http://**.

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,202 @@
+	}
+
+	switch {
+	case bytes.Contains(body, []byte("Duplicate comment")):
+		return ErrDuplicate
+	case bytes.Contains(body, []byte("Slow down")):
+		return ErrTooQuickly
+	case bytes.Contains(body, []byte("scheduled maintenance")):
+		return ErrMaintenance
```
Я бы предложил сделать эти проверки регистронезависимыми, т.е. `bytes.ToLower()` для `body`, и потом проверяем всё в нижнем регистре, т.к. не хочется пропустить кейс если сообщение просто изменит первую букву с большой на маленькую.

cravtos replied:
Готово (01c5635)

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,202 @@
+	}
+}
+
+func getPostID(productLink string) (string, error) {
```
Мне кажется, это проще брать отсюда (см. shortlink, p=id, который тебе нужен) - можно просто брать этот href и доставать id по параметру p в URL:
```html
<link rel="canonical" href="http://localhost/product/album/" />
<link rel='shortlink' href='http://localhost/?p=23' />
```

manmolecular replied:
Это тот же ID, что и comment_post_ID, но если что проверь сам

cravtos replied:
Не додумался в других местах поискать айди, спасибо.

Отказался от парсинга в пользу регулярки (для прошлого тега регулярка была бы большой, а тут коротко выходит) и теперь беру айди из шортлинка (fb82a1a). 

Казалось ещё, что с парсингом получение айдишника стабильнее будет, но скорее всего при изменениях страницы, тут что парсинг, что регулярка одинаково сломаются.

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,202 @@
+
+	timeoutDefault := 10 * time.Second
+	timeoutCur := timeoutDefault
+	timeoutLimit := 1_000 * time.Second
```
Почти 17 минут, это супер много для тайм-аут лимита. Я бы не ставил больше 30 секунд, а дальше отбрасываем эту попытку или это ревью.

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,202 @@
+		case ErrMaintenance:
+			log.Println("Site under maintenance")
+			if timeoutCur < timeoutLimit {
+				timeoutCur *= 2
```
А здесь вообще нужно увеличивать этот таймаут на два? Может, здесь динамический и не нужен, и достаточно просто подождать фиксированные 10-15 секунд и повторить опять.

cravtos replied:
15 секунд фиксированно оказалось. Изменил стандартный таймаут тут: 491afe1

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,202 @@
+	timeoutLimit := 1_000 * time.Second
+
+	reviewsDone := 0
+	for { // for reviewsDone < *reviewAmount
```
Я тут не попаду в бесконечный цикл, если например "Site under maintenance"? Т.е. количество попыток надо ограничить каким-то явным барьером, или задать с помощью CLI - например, не более 3х попыток на отправку одного ревью

manmolecular replied:
И вообще, раз у нас возможна ситуация, что приложение определяет, что комменты отправляются слишком быстро, нужно бы добавить рандомный sleep между отправкой запросов (по диапазону подумай сам, но думаю что-то в районе от 1 до 5 секунд)

cravtos replied:
Добавил флаг для указания максимального количество попыток на отправку одного ревью: 491afe1
В том же коммите при "too quickly" от сайта к дефолтному таймауту начал добавляеть 1-5 секунд.

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,202 @@
+}
+
+func getPostID(productLink string) (string, error) {
+	resp, err := http.DefaultClient.Get(productLink)
```
Здесь DefaultClient, а для POST создается свой экземпляр агента - действительно ли это необходимо? Может, унифицировать и использовать один клиент? Как вариант, сделать тогда type, типа класс, где будут содержаться те переменные и сущности, которым мы используем то там, то тут - я думаю, тут можно добавить что-то ещё, и унифицировать логику в один единый логический элемент.

Но это как вариант на подумать просто, возможно так (или иначе) будет лучше.
```go
type reviewBot struct {
    Client ...
    Host ...
    // остальные необходимые параметры
}
```

cravtos replied:
Сейчас в коде получается, что может быть два клиента: с прокси и без (прокси указывается в транспорте). Хотелось бы сделать что-то вроде конфига для клиента наверху файла (может кому нужно будет стандартные значения поменять), но в уже созданом клиенте не нашел способа изменить прокси в транспорте (напрямую нет доступа)

manmolecular reviewed (commented):
Потестил, всё работает - комменты прилетают в админку на ревью (по крайней мере, в моем варианте деплоя), так что считаем что всё норм. Если бы было авто-ревью/авто-пост, сразу бы появлялось под товаром.

Я предложил пару упрощений, где это возможно. Нужно подумать над тайм-аутами однозначно и над макс. числом попыток отправки (я думаю, не более 3х попыток). Нужно добавить таймаут между отправкой запросов в принципе, рандомный в районе от 1 до 5 секунд, чтобы не триггерить приложение на слишком быструю отправку. А вот таймаут в случае ошибки сократить - ждать 10*n секунд мне кажется слишком жестоким.

cravtos committed:
Move slices with words for review generation to global scope

cravtos committed:
Rename variables for random predefined words

cravtos committed:
Use scheme from given URL

cravtos committed:
Make error checks case insensitive

cravtos committed:
Extract post ID from shortlink

Now using regular expression instead of doing html parsing.

cravtos committed:
Add attempts counter and reduce timeout time

Add flag (-attempts) specifiyng maximal amount of attempts allowed
for sending single review. When attempts are ended, bot terminates.

Increase default timeout time to 15 seconds (starting from 15s timeouts
there is no "too quickly" messages)

cravtos committed:
Gofmt shop-fake-review.go

cravtos committed:
Add proxy support

Two new flags added:
	-proxy: Path to text file containing one proxy per line.
Program starts one bot per one proxy. If file with proxies is not
specified, proxy from $HTTP_PROXY used (default behaviour of
http.DefaultClient). Supported proxies: http, https, socks5.
	-timeout: Time given to the bot for one request.

cravtos committed:
Reuse created http.Client's

cravtos commented:
Стоит ли вынести флаги в глобальный скоп?

manmolecular commented:
> Стоит ли вынести флаги в глобальный скоп?

@Cravtos флаги лучше оставить в main, где они есть. А вот дефолтные значения флагов можно вынести, чтобы пользователь мог руками перебить что-то под себя или быстро мог увидеть, что мы юзаем за дефолт. Также в const/var. Но это на твое усмотрение, в принципе, и так, и так нормально.

На код сделаю новое ревью.

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,270 @@
+	maxAttempts := flag.Uint("attempts", 3, "Amount of attempts allowed to send single review. Zero means no limit.")
+	rating := flag.Int("rating", 5, "Rating for the reviews (1-5).")
+	proxyPath := flag.String("proxy", "", "Path to text file containing one proxy per line. "+
+		"Supported schemes: http, https, socks5. Line example: socks5://0.0.0.0:1337."+
```
В конце бы добавить пробел, иначе в хелпе строчки склеиваются

cravtos replied:
Поправил: 215506f

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,270 @@
+)
+
+func main() {
+	productLink := flag.String("product", "", "Link to the product for which to leave fake reviews for.")
```
Здесь можно указать, что также нужен протокол, или сказать, что типа "full link", любым образом

manmolecular replied:
Including scheme, etc. ...

cravtos replied:
Заставил давать фулл линк: 2834551

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,270 @@
+			Timeout: *reqTimeout,
+		}
+		wg.Add(1)
+		go startBot(wg, client, prodURL, postID, *rating, *maxAttempts, *reviewAmount, botID)
```
Я бы сюда ещё передавал используемый адрес proxy и логировал бы его вместе с BOT-ID, чтобы понимать не просто какой бот по ID не работает, но и какой конкретно прокси не работает если что-то не так.

cravtos replied:
Заменил бот айди на прокси вообще: 9aed1b8

manmolecular commented on bots/shop-fake-review/shop-fake-review.go:
```diff
@@ -0,0 +1,270 @@
+	var reviewsDone uint
+	var attempts uint
+	for { // for reviewsDone < *reviewAmount
+		log.Printf("[BOT-%d] Sending reviev\n", botID)
```
Очепятка, review

cravtos replied:
Поправил: 215506f

manmolecular reviewed (approved):
С прокси стало намного интереснее 👍🏻 
Попробовал одновременно с торовскими socks5, http, всё работает ок.

Замечания остались совсем по мелочам, фактически очепятки и оформление.

Ну а так, я бы добавил краткий комментарий над каждой функцией, какая у неё задача и что она делает, в двух словах, в формате godoc - название функи затем её описание
```go
// Fprint formats using the default formats for its operands and writes to w.
// Spaces are added between operands when neither is a string.
// It returns the number of bytes written and any write error encountered.
func Fprint(w io.Writer, a ...interface{}) (n int, err error) {
```

А так я ставлю approve.

cravtos committed:
Ask for a product link including a scheme

cravtos committed:
Correct typos

cravtos committed:
Write proxy to log prefix instead of botID

cravtos committed:
Add comments to functions

cravtos committed:
Do a minor code cleanup

	- Use postID term consistently. Earlier it was sometimes named as
prodID.
	- Add a line break to startBot function definition (too many
arguments were on one line).

cravtos committed:
Remove unused modules from go.mod

cravtos committed:
Seed RNG with current time.

Earlier seed wasn't set at all and by default integer 1 was used.
This caused many "dublicate review" errors.
